### DISCUSSION TITLE
Move constructor for ZXing::Result?

### DISCUSSION THREAD

**Discussion by duanqn** [zxing-cpp/zxing-cpp#471] [Q&A] [answered]
Hi, this just came to my mind when I was evaluating the 2.0 release. But I don't think it should block the release so I decided to start a new discussion thread.

A while ago, I was studying the performance of barcode recognition in my project. I noticed something odd. Although my PDF417 samples are the fewest among all standards I have, they have a higher frequency of appearing in the slowest images. I went on to do some profiling and did not find any obvious issue, but the program was spending a small amount of time in some places where I expected it to finish instantly.

I suspect that `ZXing::Result` is lacking a move constructor `Result(Result&&)` and this is causing the content to be copied when the `Result` object itself is copied (e.g. pushing into a vector). If this is true, then the impact is greater on my PDF417 samples because they have much longer content than other formats. I looked at the code in https://github.com/zxing-cpp/zxing-cpp/blob/master/core/src/Result.h and there was no move constructor defined. However, I'm not too confident with my speculation because I don't know if the compiler automatically generated one.

I don't have a good way to investigate this further right now, but I'd like to point out this potential issue.

@axxel for awareness.

**Comment by axxel** [comment, marked as answer]
Thanks for bringing this subject up. I checked and convinced myself that the compiler generates the default move constructor for `Result`. (If I force it to be not present, the code fails to compile all over the place). I would have had a hard time believing that a single copy vs move of the Result struct would should up on your performance measurement if that also includes a `ReadBarcode()` call.

You could add the following lines to the `Result` class and see if it makes any difference to you.
```
	Result(const Result&) = default;
	Result(Result&&) noexcept = default;
	Result& operator=(const Result&) = default;
	Result& operator=(Result&&) = default;
```

If you have a specific PDF417 image that seems to be processed slower than expected, please don't hesitate to open an issue including the image. (I will address your 'slowest.jpg' in the other comment where you mentioned it.)
